rust/scx_rustland_core/assets/bpf/main.bpf.c

```diff
@@ -479,18 +485,41 @@ s32 BPF_STRUCT_OPS(rustland_select_cpu, struct task_struct *p, s32 prev_cpu,
 	if (full_user)
 		return prev_cpu;
 
+	tctx = lookup_task_ctx(p);
+	if (!tctx)
+		return prev_cpu;
+
+	/*
+	 * If the previously used CPU is still available, keep using it to take
+	 * advantage of the cached working set.
+	 */
+	if (scx_bpf_test_and_clear_cpu_idle(prev_cpu)) {
+		tctx->dispatch_local = true;
+		return prev_cpu;
+	}
+
+	/*
+	 * If the task was directly dispatched give it a second chance to
+	 * remain on the current CPU, instead of immediately migrating it.
+	 */
+	if (tctx->dispatch_local) {
```

Comment:
And then maybe this can be renamed to clarify that it's tracking what happened the last time? Not a big difference but might make it a bit easier to follow.

Overall, yeah, other schedulers too are a lot more eager to move tasks across CPU boundaries than the kernel's. This has some benefits in terms of work conservation but at the expense of L1/2 locality. This probably is an area other schedulers can improve on too - e.g. consider overall saturation level of the system, how long the prev CPU has been busy since the waking task ran on it the last time (as a proxy for measuring cache decay), and how soon the CPU is expected to be freed up.

Reply:
Good point about renaming the variable to clarify that it's basically used to keep track of the previous direct dispatch. Will change it and clarify also the comment.

About being less aggressive with migrations, it'd be nice also to add some topology awareness to the equation, like try to stick not just on `prev_cpu`, but on a subset of CPUs that share the L2 cache with `prev_cpu`, for example. Maybe that's something that we could even add to the in-kernel idle selection logic to improve all schedulers in general?

Reply:
Yeah, teaching some topology awareness to `select_cpu_dfl()` is something on the todo list.